### Task 4.10 - Part 1

##### Table of Contents:
1.	Import the data set you exported in Exercise 4.9 into a new Jupyter notebook. This should be the merged version of your data that contains the new customers column.

2.	Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

3.	The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.
    
    o	Use the region information in this Wikipedia article to create your column (you only need to create regions, not divisions).
    
    o	Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)
4.	The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = r'C:\Users\Jay\Documents\Instacart Basket Analysis'

1.	Import the data set you exported in Exercise 4.9 into a new Jupyter notebook. This should be the merged version of your data that contains the new customers column.

In [3]:
# import prepared data
df = pd.read_pickle(os.path.join(path, '02 Data Sets', 'Prepared Data', 'ords_prods_all.pkl'))

In [4]:
df.shape

(32435059, 33)

In [5]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... frequency         frequency_flag  last_name  gender  \
0          0  ...      20.5  Non-frequent customer     Nguyen  Female   
1          0  ...      20.5  Non-frequent customer     Nguyen  Female   
2          0  ...      20.5  Non-frequent customer     Nguyen  Female   
3          0  ...      20.5  Non-frequent customer     Nguyen  Female   
4          0  ...      20.5  Non-frequent customer     Nguyen  Female   

     state age date_joined dependants family_status  income  
0  Alabama  31   2/17/2019          3       married   40423  
1  Alabama  31   2/17/2019          3       married   40423  
2  Alabama  31   2/17/2019          3       married   40423  
3  Alabama  31   2/17/2019          3       married   40423  
4  Alabama  31   2/17/2019          3       married   40423  

[5 rows x 33 columns]

2.	Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

##### As the 'first_name' column was dropped in the previous exercise, the only remaining column containing PII data is 'last_name'. The column will be dropped now, considering this information can be traced back to a particular person.

In [6]:
# drop last_name column
df = df.drop('last_name', 1)

C:\Users\Jay\AppData\Local\Temp/ipykernel_12048/1802433500.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  df = df.drop('last_name', 1)


In [7]:
df.shape

(32435059, 32)

In [8]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... spending_flag  frequency         frequency_flag  gender  \
0          0  ...   Low Spender       20.5  Non-frequent customer  Female   
1          0  ...   Low Spender       20.5  Non-frequent customer  Female   
2          0  ...   Low Spender       20.5  Non-frequent customer  Female   
3          0  ...   Low Spender       20.5  Non-frequent customer  Female   
4          0  ...   Low Spender       20.5  Non-frequent customer  Female   

     state age date_joined dependants family_status  income  
0  Alabama  31   2/17/2019          3       married   40423  
1  Alabama  31   2/17/2019          3       married   40423  
2  Alabama  31   2/17/2019          3       married   40423  
3  Alabama  31   2/17/2019          3       married   40423  
4  Alabama  31   2/17/2019          3       married   40423  

[5 rows x 32 columns]

3.	The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.

In [9]:
# Check state names and value counts
df['state'].value_counts(dropna = False)

Pennsylvania            667738
California              660428
Rhode Island            657662
Georgia                 657092
New Mexico              655188
Arizona                 654553
North Carolina          652329
Oklahoma                652275
Alaska                  649026
Minnesota               648424
Massachusetts           647020
Wyoming                 644927
Virginia                642103
Missouri                641323
Texas                   641051
Colorado                639820
Maine                   639175
North Dakota            638997
Alabama                 638650
Louisiana               638179
Kansas                  638021
Delaware                637477
South Carolina          637423
Oregon                  636966
Arkansas                636737
New York                636707
Nevada                  636697
Montana                 635838
South Dakota            634362
Illinois                633476
Hawaii                  633446
Washington              633445
Mississi

o	Use the region information in this Wikipedia article to create your column (you only need to create regions, not divisions).

In [10]:
# create regions
result = []

for value in df['state']:
    if (value == 'Maine') or (value == 'New Hampshire') or (value == 'Vermont') or (value == 'Massachusetts') or (value == 'Rhode Island') or (value == 'Connecticut') or (value == 'New York') or (value == 'Pennsylvania') or (value =='New Jersey'):
        result.append('Northeast')
    elif (value == 'Wisconsin') or (value == 'Michigan') or (value == 'Illinois') or (value == 'Indiana') or (value == 'Ohio') or (value == 'North Dakota') or (value == 'South Dakota') or (value == 'Nebraska') or (value == 'Kansas') or (value == 'Minnesota') or (value == 'Iowa') or (value == 'Missouri'):
        result.append('Midwest')
    elif (value == 'Idaho') or (value == 'Montana') or (value == 'Wyoming') or (value == 'Nevada') or (value == 'Utah') or (value == 'Colorado') or (value == 'Arizona') or (value == 'New Mexico') or (value == 'Alaska') or (value == 'Washington') or (value == 'Oregon') or (value == 'California') or (value == 'Hawaii'):
        result.append('West')
    else:
        result.append('South')

In [11]:
# print result
result

['South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'South',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'West',
 'Wes

In [12]:
# Create new column 'region'
df['region'] = result

In [13]:
# Check region counts
df['region'].value_counts(dropna = False)

South        10801982
West          8300617
Midwest       7604016
Northeast     5728444
Name: region, dtype: int64

In [14]:
df.shape

(32435059, 33)

##### Verified sum of region counts = 32,435,059

 o	Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)

In [15]:
# Create crosstab
crosstab = pd.crosstab(df['region'],df['spending_flag'],dropna = False)

In [16]:
crosstab

spending_flag  High Spender  Low Spender
region                                  
Midwest              156084      7447932
Northeast            108276      5620168
South                209878     10592104
West                 160440      8140177

In [17]:
# Midwest high spender percentage, relative to the population
Midwest = (156084/7604016)*100

In [18]:
# Northeast high spender percentage, relative to the population
Northeast = (108276/5728444)*100

In [19]:
# South high spender percentage, relative to the population
South = (209878/10801982)*100

In [20]:
# West high spender percentage, relative to the population
West = (160440/8300617)*100

In [21]:
Midwest

2.0526521774809523

In [22]:
Northeast

1.8901467833149805

In [23]:
South

1.9429582459959662

In [24]:
West

1.9328683638818658

##### The difference in number of high spenders for each region is not signifcant, when taking population into account. The four regions from highest to lowest percentile only differ by 0.16%.

4.	The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

In [25]:
# create flag for low-activity customers
df.loc[df['max_order'] < 5, 'exclusion_flag'] = 'Low-activity Customer'

In [26]:
# create flag for high-activity customers
df.loc[df['max_order'] >= 5, 'exclusion_flag'] = 'Frequent Customer'

In [27]:
# Check exclusion flag
df['exclusion_flag'].value_counts(dropna = False)

Frequent Customer        30993489
Low-activity Customer     1441570
Name: exclusion_flag, dtype: int64

In [29]:
# Create data subset including only frequent customers 
# df_freq_cust = df[df['exclusion_flag']=='Frequent Customer'] 

##### Due to above command memory error, will export df and attempt to create subset in new notebook.

In [32]:
# Export data to pkl
df.to_pickle(os.path.join(path, '02 Data Sets','Prepared Data', 'df_1.pkl'))

##### Subsets were created and exported to pkl via script titled "Test 4.10" 